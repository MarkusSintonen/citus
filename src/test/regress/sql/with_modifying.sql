-- Tests for modifying CTEs and CTEs in modifications
SET citus.next_shard_id TO 1502000;

CREATE SCHEMA with_modifying;
SET search_path TO with_modifying, public;

CREATE TABLE with_modifying.modify_table (id int, val int);
SELECT create_distributed_table('modify_table', 'id');

CREATE TABLE with_modifying.users_table (LIKE public.users_table INCLUDING ALL);
SELECT create_distributed_table('with_modifying.users_table', 'user_id');
INSERT INTO with_modifying.users_table SELECT * FROM public.users_table;

CREATE TABLE with_modifying.summary_table (id int, counter int);
SELECT create_distributed_table('summary_table', 'id');

-- basic insert query in CTE
WITH basic_insert AS (
	INSERT INTO users_table VALUES (1), (2), (3) RETURNING *
)
SELECT
	*
FROM
	basic_insert
ORDER BY
	user_id;

-- single-shard UPDATE in CTE
WITH basic_update AS (
	UPDATE users_table SET value_3=41 WHERE user_id=1 RETURNING *
)
SELECT
	*
FROM
	basic_update
ORDER BY
	user_id,
	time
LIMIT 10;

-- multi-shard UPDATE in CTE
WITH basic_update AS (
	UPDATE users_table SET value_3=42 WHERE value_2=1 RETURNING *
)
SELECT
	*
FROM
	basic_update
ORDER BY
	user_id,
	time
LIMIT 10;

-- single-shard DELETE in CTE
WITH basic_delete AS (
	DELETE FROM users_table WHERE user_id=6 RETURNING *
)
SELECT
	*
FROM
	basic_delete
ORDER BY
	user_id,
	time
LIMIT 10;

-- multi-shard DELETE in CTE
WITH basic_delete AS (
	DELETE FROM users_table WHERE value_3=41 RETURNING *
)
SELECT
	*
FROM
	basic_delete
ORDER BY
	user_id,
	time
LIMIT 10;

-- INSERT...SELECT query in CTE
WITH copy_table AS (
	INSERT INTO users_table SELECT * FROM users_table WHERE user_id = 0 OR user_id = 3 RETURNING *
)
SELECT
	*
FROM
	copy_table
ORDER BY
	user_id,
	time
LIMIT 10;

-- CTEs prior to INSERT...SELECT via the coordinator should work
WITH cte AS (
	SELECT user_id FROM users_table WHERE value_2 IN (1, 2)
)
INSERT INTO modify_table (SELECT * FROM cte);


WITH cte_1 AS (
	SELECT user_id, value_2 FROM users_table WHERE value_2 IN (1, 2, 3, 4)
),
cte_2 AS (
	SELECT user_id, value_2 FROM users_table WHERE value_2 IN (3, 4, 5, 6)
)
INSERT INTO modify_table (SELECT cte_1.user_id FROM cte_1 join cte_2 on cte_1.value_2=cte_2.value_2);


-- even if this is an INSERT...SELECT, the CTE is under SELECT
WITH cte AS (
	SELECT user_id, value_2 FROM users_table WHERE value_2 IN (1, 2)
)
INSERT INTO modify_table (SELECT (SELECT value_2 FROM cte GROUP BY value_2));


-- CTEs prior to any other modification should error out
WITH cte AS (
	SELECT value_2 FROM users_table WHERE user_id IN (1, 2, 3)
)
DELETE FROM modify_table WHERE id IN (SELECT value_2 FROM cte);


WITH cte AS (
	SELECT value_2 FROM users_table WHERE user_id IN (1, 2, 3)
)
UPDATE modify_table SET val=-1 WHERE val IN (SELECT * FROM cte);


WITH cte AS (
	WITH basic AS (
		SELECT value_2 FROM users_table WHERE user_id IN (1, 2, 3)
	)
	INSERT INTO modify_table (SELECT * FROM basic) RETURNING *
)
UPDATE modify_table SET val=-2 WHERE id IN (SELECT id FROM cte);


WITH cte AS (
	WITH basic AS (
		SELECT * FROM events_table WHERE event_type = 5
	),
	basic_2 AS (
		SELECT user_id FROM users_table
	)
	INSERT INTO modify_table (SELECT user_id FROM events_table) RETURNING *
)
SELECT * FROM cte;

WITH user_data AS (
	SELECT user_id, value_2 FROM users_table
)
INSERT INTO modify_table SELECT * FROM user_data;

WITH raw_data AS (
	DELETE FROM modify_table RETURNING *
)
INSERT INTO summary_table SELECT id, COUNT(*) AS counter FROM raw_data GROUP BY id;

SELECT * FROM summary_table ORDER BY id;
SELECT COUNT(*) FROM modify_table;

INSERT INTO modify_table VALUES (1,1), (2, 2), (3,3);

WITH raw_data AS (
	DELETE FROM modify_table RETURNING *
)
INSERT INTO summary_table SELECT id, COUNT(*) AS counter FROM raw_data GROUP BY id;

SELECT * FROM summary_table ORDER BY id, counter;
SELECT COUNT(*) FROM modify_table;

-- merge rows in the summary_table
WITH raw_data AS (
	DELETE FROM summary_table RETURNING *
)
INSERT INTO summary_table SELECT id, SUM(counter) AS counter FROM raw_data GROUP BY id;

SELECT * FROM summary_table ORDER BY id;

-- check modifiying CTEs inside a transaction
INSERT INTO modify_table VALUES (1,1), (2, 2), (3,3);

BEGIN;
WITH raw_data AS (
	DELETE FROM summary_table RETURNING *
)
INSERT INTO summary_table SELECT id, SUM(counter) AS counter FROM raw_data GROUP BY id;
ROLLBACK;

SELECT COUNT(*) FROM modify_table;
SELECT * FROM summary_table ORDER BY id, counter;

CREATE TABLE with_modifying.anchor_table (id int);
SELECT create_distributed_table('anchor_table', 'id');

INSERT INTO anchor_table VALUES (1), (2);

WITH raw_data AS (
	DELETE FROM modify_table RETURNING *
),
anchor_data AS (
	DELETE FROM anchor_table RETURNING *
)
INSERT INTO
	summary_table
SELECT raw_data.id, COUNT(*) AS counter FROM raw_data, anchor_data
	WHERE raw_data.id = anchor_data.id GROUP BY raw_data.id;

SELECT COUNT(*) FROM modify_table;
SELECT * FROM summary_table ORDER BY id, counter;

WITH added_data AS (
	INSERT INTO modify_table VALUES (1,1), (1,3), (2, 2), (3,3) RETURNING *
),
raw_data AS (
	DELETE FROM modify_table WHERE id = 1 AND val = (SELECT MAX(val) FROM added_data) RETURNING *
)
INSERT INTO summary_table SELECT id, COUNT(*) AS counter FROM raw_data GROUP BY id;

SELECT COUNT(*) FROM modify_table;
SELECT * FROM summary_table ORDER BY id, counter;

DROP SCHEMA with_modifying CASCADE;
