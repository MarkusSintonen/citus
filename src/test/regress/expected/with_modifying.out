-- Tests for modifying CTEs and CTEs in modifications
SET citus.next_shard_id TO 1502000;
CREATE SCHEMA with_modifying;
SET search_path TO with_modifying, public;
CREATE TABLE with_modifying.modify_table (id int, val int);
SELECT create_distributed_table('modify_table', 'id');
 create_distributed_table 
--------------------------
 
(1 row)

CREATE TABLE with_modifying.users_table (LIKE public.users_table INCLUDING ALL);
SELECT create_distributed_table('with_modifying.users_table', 'user_id');
 create_distributed_table 
--------------------------
 
(1 row)

INSERT INTO with_modifying.users_table SELECT * FROM public.users_table;
-- basic insert query in CTE
WITH basic_insert AS (
	INSERT INTO users_table VALUES (1), (2), (3) RETURNING *
)
SELECT
	*
FROM
	basic_insert;
 user_id | time | value_1 | value_2 | value_3 | value_4 
---------+------+---------+---------+---------+---------
       1 |      |         |         |         |        
       3 |      |         |         |         |        
       2 |      |         |         |         |        
(3 rows)

-- single-shard UPDATE in CTE
WITH basic_update AS (
	UPDATE users_table SET value_3=41 WHERE user_id=1 RETURNING *
)
SELECT
	*
FROM
	basic_update
ORDER BY
	user_id,
	time;
 user_id |              time               | value_1 | value_2 | value_3 | value_4 
---------+---------------------------------+---------+---------+---------+---------
       1 | Wed Nov 22 22:51:43.132261 2017 |       4 |       0 |      41 |        
       1 | Thu Nov 23 03:32:50.803031 2017 |       3 |       2 |      41 |        
       1 | Thu Nov 23 09:26:42.145043 2017 |       1 |       3 |      41 |        
       1 | Thu Nov 23 11:11:24.40789 2017  |       3 |       4 |      41 |        
       1 | Thu Nov 23 11:44:57.515981 2017 |       4 |       3 |      41 |        
       1 | Thu Nov 23 17:23:03.441394 2017 |       5 |       4 |      41 |        
       1 | Thu Nov 23 17:30:34.635085 2017 |       3 |       4 |      41 |        
       1 |                                 |         |         |      41 |        
       1 |                                 |         |         |      41 |        
(9 rows)

-- multi-shard UPDATE in CTE
WITH basic_update AS (
	UPDATE users_table SET value_3=42 WHERE value_2=1 RETURNING *
)
SELECT
	*
FROM
	basic_update
ORDER BY
	user_id,
	time;
 user_id |              time               | value_1 | value_2 | value_3 | value_4 
---------+---------------------------------+---------+---------+---------+---------
       2 | Thu Nov 23 06:50:30.797805 2017 |       1 |       1 |      42 |        
       2 | Thu Nov 23 06:56:38.46819 2017  |       0 |       1 |      42 |        
       2 | Thu Nov 23 13:52:54.83829 2017  |       3 |       1 |      42 |        
       3 | Wed Nov 22 18:43:51.450263 2017 |       1 |       1 |      42 |        
       3 | Thu Nov 23 00:15:45.610845 2017 |       1 |       1 |      42 |        
       4 | Wed Nov 22 23:59:46.493416 2017 |       3 |       1 |      42 |        
       4 | Thu Nov 23 01:55:21.824618 2017 |       3 |       1 |      42 |        
       4 | Thu Nov 23 06:50:08.101207 2017 |       2 |       1 |      42 |        
       4 | Thu Nov 23 07:09:37.382372 2017 |       4 |       1 |      42 |        
       4 | Thu Nov 23 08:38:45.877401 2017 |       4 |       1 |      42 |        
       4 | Thu Nov 23 11:22:36.283022 2017 |       4 |       1 |      42 |        
       4 | Thu Nov 23 12:59:23.28005 2017  |       4 |       1 |      42 |        
       4 | Thu Nov 23 13:56:39.967304 2017 |       0 |       1 |      42 |        
       4 | Thu Nov 23 14:57:36.278713 2017 |       3 |       1 |      42 |        
       5 | Wed Nov 22 22:31:47.62577 2017  |       3 |       1 |      42 |        
       5 | Thu Nov 23 06:36:11.898621 2017 |       5 |       1 |      42 |        
       5 | Thu Nov 23 09:17:14.659809 2017 |       3 |       1 |      42 |        
       5 | Thu Nov 23 10:46:56.036776 2017 |       3 |       1 |      42 |        
       5 | Thu Nov 23 14:29:02.557934 2017 |       2 |       1 |      42 |        
       6 | Wed Nov 22 20:15:53.317797 2017 |       1 |       1 |      42 |        
       6 | Thu Nov 23 00:07:11.068353 2017 |       1 |       1 |      42 |        
(21 rows)

-- single-shard DELETE in CTE
WITH basic_delete AS (
	DELETE FROM users_table WHERE user_id=6 RETURNING *
)
SELECT
	*
FROM
	basic_delete;
 user_id |              time               | value_1 | value_2 | value_3 | value_4 
---------+---------------------------------+---------+---------+---------+---------
       6 | Thu Nov 23 00:09:44.19812 2017  |       5 |       2 |       0 |        
       6 | Thu Nov 23 14:43:18.024104 2017 |       3 |       2 |       5 |        
       6 | Wed Nov 22 20:15:53.317797 2017 |       1 |       1 |      42 |        
       6 | Thu Nov 23 13:51:16.92838 2017  |       0 |       4 |       2 |        
       6 | Thu Nov 23 00:07:11.068353 2017 |       1 |       1 |      42 |        
       6 | Thu Nov 23 01:14:55.769581 2017 |       0 |       0 |       5 |        
       6 | Thu Nov 23 11:08:04.244582 2017 |       2 |       3 |       2 |        
       6 | Thu Nov 23 01:13:50.526322 2017 |       2 |       4 |       1 |        
       6 | Thu Nov 23 10:22:11.02918 2017  |       5 |       0 |       5 |        
       6 | Wed Nov 22 23:01:24.82289 2017  |       2 |       4 |       1 |        
       6 |                                 |         |         |         |        
(11 rows)

-- multi-shard DELETE in CTE
WITH basic_delete AS (
	DELETE FROM users_table WHERE value_3=41 RETURNING *
)
SELECT
	*
FROM
	basic_delete;
 user_id |              time               | value_1 | value_2 | value_3 | value_4 
---------+---------------------------------+---------+---------+---------+---------
       1 | Thu Nov 23 17:30:34.635085 2017 |       3 |       4 |      41 |        
       1 | Thu Nov 23 11:11:24.40789 2017  |       3 |       4 |      41 |        
       1 | Wed Nov 22 22:51:43.132261 2017 |       4 |       0 |      41 |        
       1 | Thu Nov 23 09:26:42.145043 2017 |       1 |       3 |      41 |        
       1 | Thu Nov 23 11:44:57.515981 2017 |       4 |       3 |      41 |        
       1 | Thu Nov 23 17:23:03.441394 2017 |       5 |       4 |      41 |        
       1 | Thu Nov 23 03:32:50.803031 2017 |       3 |       2 |      41 |        
       1 |                                 |         |         |      41 |        
       1 |                                 |         |         |      41 |        
(9 rows)

-- INSERT...SELECT query in CTE
WITH copy_table AS (
	INSERT INTO users_table SELECT * FROM users_table WHERE user_id = 0 OR user_id = 3 RETURNING *
)
SELECT
	*
FROM
	copy_table;
 user_id |              time               | value_1 | value_2 | value_3 | value_4 
---------+---------------------------------+---------+---------+---------+---------
       3 | Thu Nov 23 11:18:53.114408 2017 |       2 |       2 |       0 |        
       3 | Wed Nov 22 18:43:51.450263 2017 |       1 |       1 |      42 |        
       3 | Thu Nov 23 00:15:45.610845 2017 |       1 |       1 |      42 |        
       3 | Thu Nov 23 03:23:24.702501 2017 |       1 |       2 |       5 |        
       3 | Thu Nov 23 11:31:17.403189 2017 |       4 |       5 |       3 |        
       3 | Thu Nov 23 11:41:21.157066 2017 |       3 |       5 |       3 |        
       3 | Thu Nov 23 17:10:35.959913 2017 |       4 |       3 |       1 |        
       3 | Wed Nov 22 23:24:32.080584 2017 |       3 |       2 |       5 |        
       3 | Thu Nov 23 05:01:44.885505 2017 |       3 |       5 |       4 |        
       3 | Wed Nov 22 20:43:31.008625 2017 |       1 |       3 |       2 |        
       3 | Thu Nov 23 12:56:49.29191 2017  |       0 |       5 |       1 |        
       3 | Thu Nov 23 17:18:51.048758 2017 |       1 |       5 |       5 |        
       3 | Thu Nov 23 09:57:41.540228 2017 |       2 |       2 |       3 |        
       3 | Thu Nov 23 04:01:08.04806 2017  |       5 |       5 |       3 |        
       3 | Thu Nov 23 03:52:32.008895 2017 |       4 |       2 |       0 |        
       3 | Thu Nov 23 12:43:49.261057 2017 |       4 |       3 |       1 |        
       3 | Thu Nov 23 06:20:05.854857 2017 |       1 |       4 |       2 |        
       3 |                                 |         |         |         |        
       3 |                                 |         |         |         |        
(19 rows)

-- CTEs prior to INSERT...SELECT via the coordinator should work
WITH cte AS (
	SELECT user_id FROM users_table WHERE value_2 IN (1, 2)
)
INSERT INTO modify_table (SELECT * FROM cte);
WITH cte_1 AS (
	SELECT user_id, value_2 FROM users_table WHERE value_2 IN (1, 2, 3, 4)
),
cte_2 AS (
	SELECT user_id, value_2 FROM users_table WHERE value_2 IN (3, 4, 5, 6)
)
INSERT INTO modify_table (SELECT cte_1.user_id FROM cte_1 join cte_2 on cte_1.value_2=cte_2.value_2);
-- even if this is an INSERT...SELECT, the CTE is under SELECT
WITH cte AS (
	SELECT user_id, value_2 FROM users_table WHERE value_2 IN (1, 2)
)
INSERT INTO modify_table (SELECT (SELECT value_2 FROM cte GROUP BY value_2));
WARNING:  more than one row returned by a subquery used as an expression
CONTEXT:  while executing command on localhost:57638
ERROR:  could not receive query results
-- CTEs prior to any other modification should error out
WITH cte AS (
	SELECT value_2 FROM users_table WHERE user_id IN (1, 2, 3)
)
DELETE FROM modify_table WHERE id IN (SELECT value_2 FROM cte);
ERROR:  common table expressions are not supported in distributed modifications
WITH cte AS (
	SELECT value_2 FROM users_table WHERE user_id IN (1, 2, 3)
)
UPDATE modify_table SET val=-1 WHERE val IN (SELECT * FROM cte);
ERROR:  common table expressions are not supported in distributed modifications
WITH cte AS (
	WITH basic AS (
		SELECT value_2 FROM users_table WHERE user_id IN (1, 2, 3)
	)
	INSERT INTO modify_table (SELECT * FROM basic) RETURNING *
)
UPDATE modify_table SET val=-2 WHERE id IN (SELECT id FROM cte);
ERROR:  common table expressions are not supported in distributed modifications
WITH cte AS (
	WITH basic AS (
		SELECT * FROM events_table WHERE event_type = 5
	),
	basic_2 AS (
		SELECT user_id FROM users_table
	)
	INSERT INTO modify_table (SELECT user_id FROM events_table) RETURNING *
)
SELECT * FROM cte;
ERROR:  RETURNING is not supported in INSERT ... SELECT via coordinator
DROP SCHEMA with_modifying CASCADE;
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to table modify_table
drop cascades to table users_table
